File: Algorithm-Swift/Algorithm-Swift/Inflearn/83_복면산.swift
//
//  83_복면산.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/09/12.
//

import Foundation

var arr83 = Array(repeating: 0, count: 101)

func send() -> Int {
    return arr83[6] * 1000 + arr83[1] * 100 + arr83[3] * 10 + arr83[0]
}

func more() -> Int {
    return arr83[2] * 1000 + arr83[4] * 100 + arr83[5] * 10 + arr83[1]
}

func money() -> Int {
    return arr83[2] * 10000 + arr83[4] * 1000 + arr83[3] * 100 + arr83[1] * 10 + arr83[7]
}

func dfs83(l: Int, ch: inout [Int]) {
    if l == 8 {
        if send() + more() == money() {
            if arr83[6] == 0 || arr83[2] == 0 { return }
            print("  ", arr83[6], arr83[1], arr83[3], arr83[0])
            print("+ ", arr83[2], arr83[4], arr83[5], arr83[1])
            print("--------------")
            print(arr83[2], arr83[4], arr83[3], arr83[1], arr83[7])
        }
    } else {
        (0..<10).forEach { i in
            if ch[i] == 0 {
                arr83[l] = i
                ch[i] = 1
                dfs83(l: l + 1, ch: &ch)
                ch[i] = 0
            }
        }
    }
}

func solution83() {
    var ch = Array(repeating: 0, count: 101)
    dfs83(l: 0, ch: &ch)
}

File: Algorithm-CPP/83_복면산.cpp
#include <iostream>
using namespace std;
int ch[101], a[101];

int send() {
    return a[6] * 1000 + a[1] * 100 + a[3] * 10 + a[0];
}

int more() {
    return a[2] * 1000 + a[4] * 100 + a[5] * 10 + a[1];
}

int money() {
    return a[2] * 10000 + a[4] * 1000 + a[3] * 100 + a[1] * 10 + a[7];
}

void DFS(int l) {
    if(l == 8) {
        if(send() + more() == money()) {
            if(a[2] == 0 || a[6] == 0) return;
            printf("   %d %d %d %d\n", a[6], a[1], a[3], a[0]);
            printf("+  %d %d %d %d\n", a[2], a[4], a[5], a[1]);
            printf("--------------\n");
            printf(" %d %d %d %d %d\n", a[2], a[4], a[3], a[1], a[7]);
        }
    } else {
        for(int i = 0; i < 10; i++) {
            if(ch[i] == 0) {
                a[l] = i;
                ch[i] = 1;
                DFS(l + 1);
                ch[i] = 0;
            }
        }
    }
}

int main() {
    DFS(0);
    return 0;
}